Add some unit tests

takikawa opened the pull request:
This PR adds some initial unit tests. It's not quite done, in that I'm having trouble getting the tests for data packets to work (commented out), but I'd like to have a first version reviewed and merged rather than stay stuck on it.

This also adjusts the polling period for the process node.

xray7224 commented on ipfix/test_ipfix.py:
```diff
@@ -0,0 +1,135 @@
+                self.assert_equal(ipfix.version, 10)
+                # check that there's an IPFIX template
+                self.assert_equal(ipfix[Template].templateID, 256)
+            except:
```
It's usually better to catch the specific errors you're looking for. Seems to me you're looking for both the key error (i.e. you can't find the part of the packet you expected to and thus it's thrown an error) and assertion errors (i.e. the packet was valid but not what you expected). I'm thinking something like this:
```python
try:
     ip = packet[IP]
except KeyError:
    # Don't know if this is the correct exception for packet[IPFIX] or others to throw exception
    self.logger.error(ppp("Invalid packet:", packet))
    raise
except AssertionError:
    self.logger.error(ppp("Unexpected packet:", packet))
    raise
```

xray7224 commented on ipfix/test_ipfix.py:
```diff
@@ -0,0 +1,135 @@
+#   $ make test TEST=test_ipfix EXTERN_TESTS=<path-to-plugin>/ipfix \
+#               EXTERN_PLUGIN=/usr/lib/vpp_plugins
+#
+# To help debug the tests, you can supply a verbosity flag, e.g., "V=1"
```
I think this probably should go in the module level docstring. This is more of an aesthetic change but it's more pythonic to do that in my opinion.

Something like this:
```python
#!/usr/bin/env python
"""
Unit tests for IPFIX plugin

...

To help debug the tests, you can supply a verbosity flag e.g., V=1
"""
```
More info here: https://www.python.org/dev/peps/pep-0257/

takikawa commented:
Thanks for the comments. I just added some more commits that fixes some issues in the plugin that made it difficult to unit test. I'll push some more commits to address your comments too.

takikawa commented:
Ok added some commits addressing your comments on the python code. @xray7224 if you have a chance to take a look at the new changes I had to make in the plugins for the data unit test, that'd be great. If it looks ok after then I'll merge. Thanks.

takikawa commented:
One problem I realized with this PR is that changing the feature arc to `ip4-output` like I did rather than `ip4-unicast` switches it to doing outbound IPFIX probing and not inbound. Which doesn't matter that much for our tests, but it's not doing the same thing as Snabb in a sense. I couldn't get the unit tests to work with `ip4-unicast` though...

xray7224 commented on ipfix/test_ipfix.py:
```diff
@@ -0,0 +1,161 @@
+            i.admin_up()
+            i.config_ip4()
+            i.resolve_arp()
+
```
Another small nit, I'd have probably created a `AbstractIPFIXTestCase` class or something since the `setUpClass` and `setUp` methods seem to be shard amongst them (as well as some other stuff I mention below).

xray7224 commented on ipfix/test_ipfix.py:
```diff
@@ -0,0 +1,161 @@
+        capture0 = self.pg0.get_capture(timeout = 5, expected_count = 1)
+        self.verify_capture(self.pg0, self.pg1, capture0)
+
+    def verify_capture(self, collector_if, exporter_if, capture):
```
I wonder if it makes sense if it's inheriting from a base class to have a method for verifying each packet and then in the `verify_template` method you can then do like

```python
def verify_template(self, collector_if, exporter_if, capture):
    for packet in capture:
        try:
            self.verify_packet(packet)
            self.assert_equal(packet[IP][UDP][IPFIX][Template].templateID, 256)
        except IndexError:
            # stuff here
        except AssertionError:
            # stuff here
```

Another option could be to have a common verification method which just takes a `template=False` argument, or of course what it is now is fine too :)

WYDT?
